components/tree/__tests__/util.test.js

```diff
@@ -32,4 +32,51 @@ describe('Tree util', () => {
     const target = ['0-0-0', '0-0-1', '0-1', '0-2', '0-2-0', '0-2-0-0', '0-2-0-1'];
     expect(keys.sort()).toEqual(target.sort());
   });
+
+  it('calc range keys by treeData', () => {
+    const treeData = [
+      {
+        key: '0-0-0',
+        title: 'Folder',
+        children: [
+          {
+            title: 'Folder2',
+            key: '0-0-1',
+            children: [
+              {
+                title: 'File',
+                key: '0-0-2',
+                isLeaf: true,
+              },
+            ],
+          },
+        ],
+      },
+      {
+        key: '0-0-3',
+        title: 'Folder',
+        children: [
+          {
+            title: 'File',
+            key: '0-0-4',
+            isLeaf: true,
+          },
+          {
+            title: 'File',
+            key: '0-0-5',
+            isLeaf: true,
+          },
+          {
+            title: 'File',
+            key: '0-0-6',
+            isLeaf: true,
+          },
+        ],
+      },
+    ];
+
+    const keys = getFullKeyListByTreeData(treeData);
```

Comment:
You should test Tree component behavior instead of testing inner function.

Reply:
I am sorry, this is my first contribution for ant-design, I have see documents but I still don't understand how to write this test, can you help me ?

Reply:
should I white test case in directory.test.js ? like this:
```

it('defaultExpandAll when use treeData', () => {
    const treeData = [
      {
        key: '0-0-0',
        title: 'Folder',
        children: [
          {
            title: 'Folder2',
            key: '0-0-1',
            children: [
              {
                title: 'File',
                key: '0-0-2',
                isLeaf: true,
              },
            ],
          },
        ],
      },
      {
        key: '0-0-3',
        title: 'Folder',
        children: [
          {
            title: 'File',
            key: '0-0-4',
            isLeaf: true,
          },
          {
            title: 'File',
            key: '0-0-5',
            isLeaf: true,
          },
          {
            title: 'File',
            key: '0-0-6',
            isLeaf: true,
          },
        ],
      },
    ];
    const wrapper = render(createTree({ defaultExpandAll: true, treeData }));
    expect(wrapper).toMatchSnapshot();
  });
```

Reply:
Just simulate the exact problem in test case, like test case in https://github.com/ant-design/ant-design/blob/master/components/tree/__tests__/directory.test.js or https://github.com/ant-design/ant-design/blob/fb5d709ccdd8d1164b7b59982ea5ab9b9848c4c4/components/menu/__tests__/index.test.js#L327

Reply:
so this is right ?  I don't sure...
```
it('DirectoryTree should expend all when use treeData and defaultExpandAll is true', () => {
    const treeData = [
      {
        key: '0-0-0',
        title: 'Folder',
        children: [
          {
            title: 'Folder2',
            key: '0-0-1',
            children: [
              {
                title: 'File',
                key: '0-0-2',
                isLeaf: true,
              },
            ],
          },
        ],
      },
    ];
    const wrapper = render(createTree({ defaultExpandAll: true, treeData }));
    expect(wrapper).toMatchSnapshot();
  });
```

Reply:
If this test case can reflect the problem you meet, then it is right.